langchain/src/vectorstores/clickhouse.ts

```diff
@@ -0,0 +1,333 @@
+   */
+  static async fromDocuments(
+    docs: Document[],
+    embeddings: Embeddings,
+    args: ClickHouseLibArgs
+  ): Promise<ClickHouseStore> {
+    const instance = new this(embeddings, args);
+    await instance.addDocuments(docs);
+    return instance;
+  }
+
+  /**
+   * Static method to create an instance of ClickHouseStore from an existing
+   * index.
+   * @param embeddings The embeddings to use.
+   * @param args The arguments for the ClickHouseStore.
+   * @returns Promise that resolves with a new instance of ClickHouseStore.
+   */
+  static async fromExistingIndex(
+    embeddings: Embeddings,
+    args: ClickHouseLibArgs
+  ): Promise<ClickHouseStore> {
+    const instance = new this(embeddings, args);
+
+    await instance.initialize();
+    return instance;
+  }
+
+  /**
+   * Method to initialize the ClickHouse database.
+   * @param dimension Optional dimension of the vectors.
+   * @returns Promise that resolves when the database has been initialized.
+   */
+  private async initialize(dimension?: number): Promise<void> {
+    const dim = dimension ?? (await this.embeddings.embedQuery("test")).length;
+
+    const indexParamStr = this.indexParam
+      ? Object.entries(this.indexParam)
+          .map(([key, value]) => `'${key}', ${value}`)
+          .join(", ")
+      : "";
+
+    const query = `
+    CREATE TABLE IF NOT EXISTS ${this.database}.${this.table}(
+      ${this.columnMap.id} Nullable(String),
+      ${this.columnMap.document} Nullable(String),
+      ${this.columnMap.embedding} Array(Float32),
+      ${this.columnMap.metadata} JSON,
+      ${this.columnMap.uuid} UUID DEFAULT generateUUIDv4(),
+      CONSTRAINT cons_vec_len CHECK length(${this.columnMap.embedding}) = ${dim},
+      INDEX vec_idx ${this.columnMap.embedding} TYPE ${this.indexType}(${indexParamStr}) GRANULARITY 1000
+    ) ENGINE =  MergeTree ORDER BY ${this.columnMap.uuid} SETTINGS index_granularity = 8192;`;
+
+    await this.client.exec({
+      query,
+      clickhouse_settings: {
+        allow_experimental_object_type: 1,
+        allow_experimental_annoy_index: 1,
+      },
+    });
+    this.isInitialized = true;
+  }
+
+  /**
+   * Method to build an SQL query for inserting vectors and documents into
+   * the ClickHouse database.
+   * @param vectors The vectors to insert.
+   * @param documents The documents to insert.
+   * @returns The SQL query string.
+   */
+  private buildInsertQuery(vectors: number[][], documents: Document[]): string {
+    const columnsStr = Object.values(
+      Object.fromEntries(
+        Object.entries(this.columnMap).filter(
+          ([key]) => key !== this.columnMap.uuid
+        )
+      )
+    ).join(", ");
+
+    const data: string[] = [];
+    for (let i = 0; i < vectors.length; i += 1) {
+      const vector = vectors[i];
+      const document = documents[i];
+      const item = [
+        `'${uuid.v4()}'`,
+        `'${this.escapeString(document.pageContent)}'`,
```

Comment:
I think some risk of SQL injection here?

Reply:
Push a new commit to resolve the risk